### ISSUE TITLE
Question on order of callbacks within a module with registration on parents verses child nodes

### ISSUE THREAD

**Issue by bbergquist0930** [sysrepo/sysrepo#2987]
I have a question on the order of callback for change subscriptions in the same module.   Take a YANG model that corresponds to this XML 

```
<vc xmlns="urn:canogaperkins:tsx:vcs" xmlns:t="urn:canogaperkins:tsx:types">
        <internal-vc-id>2</internal-vc-id>
        <uni-endpoints>
          <port>t:Uni-1</port>
          <internal-endpoint-id>2</internal-endpoint-id>
          <mac-lookup-needed>false</mac-lookup-needed>
          <portid/>
        </uni-endpoints>
        <egress-flows>
          <egress-flow-id>104</egress-flow-id>
          <destinations>
            <destination-endpoint-id>1</destination-endpoint-id>
            <destination-mac-address>00:40:2A:A1:CE:2E</destination-mac-address>
          </destinations>
        </egress-flows>
      </vc>
```

where `vc`, `uni-endpoints`, `egress-flows`, `egress-flows/destinations` are all list containers.  

If I have a change callbacks registered on 
* vc/vc-internal-id
* vc/uni-endpoints/*
* vc/egress-flows/*
* vc/egress-flows/destinations/*
all with the same priority, is there any order on the subscription callbacks?   Could I see the callback for `vc/egress-flows/destinations/*` before the callback for `vc/egress-flows/*` for example?

Basically can you state the order of callbacks within a module in terms of parent node callbacks and child node callbacks.

This question is for sysrepo 2.x?

**Comment by michalvasko** [comment]
Subscriptions for a single module use the same SHM segment. In case there are several subscribers with the same priority, the event originator simply writes their number in the SHM segment and waits until they all handle the event meaning it enforces no order and has no control over it.

So it depends on the subscribers. If all the subscribers are handled by a single subscription structure in a single thread, the callbacks are called in the same order the subscriptions were made. But if the subscriptions are handled by several threads, this order is kept only for each thread, one thread's subscriptions are arbitrarily intertwined with other threads. In short, if you care about the order, you best use the priorities. That is what they are there for.

**Comment by bbergquist0930** [comment]
Thank you very much!  

Interesting how the universe works in that I have a piece of code with the same priorities and over the last year, Netconf requests through Netopeer2-server to Sysrepo to the application with the callback always called the callbacks in the order that they were registered.  Manually doing a change through `sysrepocfg` triggered a different callback order.   

The bug was in the code always, just the cosmic rays never lined up just right to trigger it.

Can I make one suggestion in that in the documentation there is for `sr_module_change_subscribe`
>[in] | priority | Specifies the order in which the callbacks (within module) will be called.

could you maybe add something like `from higher priority to lower priority` just to make it clear which way priorities are treated.   I did do some test code and I know that it is from higher to lower, but this might help the next person that reads the documentation.

Again thank you for the response and the issue can be closed or I will close it later today.

**Comment by michalvasko** [comment]
Priorities clarified.
